### PR TITLE
CAN Bootloader

### PR DESCRIPTION
**Pull request by gtaharaedmonds**
### Summary

Create a CAN bootloader for STM32F412 targets so we can update their firmware over the CAN bus. A bootloader is a separate program running on the MCU that (unsurprisingly) handles:
- Booting: Can start the main program (usually called the "app", not to be confused with our app-level abstraction layer)
- Loading: Can update its app code

When you reset the microcontroller, instead of running the usual program it now runs the bootloader. Its then up to the bootloader to decide whether or not to start the app or update its code. A microcontroller is able to overwrite its own flash: So its only a problem of telling the microcontroller what data to write at what location.

### Changelist 

- Create STM32F4 CAN bootloader in `firmware/boot` directory. The bootloader handles updating the application code, and uses STM's cyclic redundancy check peripheral to verify that the application was programmed correctly. Also, it broadcasts its application state at 10Hz over CAN while in the bootloader.
- Created linker scripts for the app and bootloader in `firmware/linker` to place the programs in their respective locations in memory.
- Add Python script to parse and send the bootloader a binary over CAN. Unfortunately this has to be a separate repo, since WSL has no CAN support whatsoever. This can be used to update or erase boards' applications.
- Add shared `hw` drivers for FLASH memory and cyclic redundancy check peripherals

### Testing Done

- [x] Testing on all boards

### Checklist
*Please change `[ ]` to `[x]` when you are ready.*
- [x] I have read and followed the code conventions detailed in [README.md](../README.md) (*This will save time for both you and the reviewer!*).
- [x] If this pull request is longer then **500** lines, I have provided *explicit* justification in the summary above explaining why I *cannot* break this up into multiple pull requests (*Small PR's are faster and less painful for everyone involved!*).

### CODE REVIEW

**Comment by Lucien950** [comment]
image
what's up with all these new targets?

**Review by Lucien950 — CHANGES_REQUESTED** [review]
I couldn't highlight it in the review section cause it's a submodule, but can you also add the requirements.txt to the requiremnets.txt in the environment folder, and Pipenv?

**Comment by Lucien950** [comment]
Also python command documentation pls because I want to have a Clion config for Python I can add here.

**Review by will-chaba — APPROVED** [review]
Fix tests and maybe add some newlines at end of files

File: can_bus/json/shared_enum.json [hunk 1]
```diff
@@ -58,5 +58,10 @@
 		"RANGE_CHECK_OK": 0,
 		"RANGE_CHECK_UNDERFLOW": 1,
 		"RANGE_CHECK_OVERFLOW": 2
-	}
+	},
+    "BootStatus": {
+        "BOOT_STATUS_NO_APP": 0,
+        "BOOT_STATUS_APP_INVALID": 1,
+        "BOOT_STATUS_APP_VALID": 2
```
Comment (Lucien950) [inline can_bus/json/shared_enum.json:65]: probably better to make happy state 0, and natural numbers error codes.

**Pull request merged**